[plugins] Add hooks to the dnf_goal_depsolve function

jrohel opened the pull request:
Added hooks:
PLUGIN_HOOK_ID_CONTEXT_PRE_GOAL_SOLVE
   Called just before the goal is resolved. The place where the plugin
   have a chance to change the goal.

PLUGIN_HOOK_ID_CONTEXT_GOAL_SOLVED
   Called shortly after the goal is resolved.

Added functions for accessing DnfPluginHookData delivered to plugins:
HyGoal hookContextGoalGetGoal(DnfPluginHookData * data);
DnfGoalActions * hookContextGoalGetActions(DnfPluginHookData * data);

Warning:
There was a missing context in the dnf_goal_depsolve function.
The code adds a pointer to DnfContext into the sack. The pointer is set
in the "dnf_context_setup_sack_with_flags" function. However, PackageKit
has it's own implementation of sack setup. So pointer to the context
will not be set in sack and the hooks will not be called.

In order for new hooks to be called with PackageKit, PackageKit must be
modified. -> Add a call to "dnf_sack_set_context (bag, job_data-> context);"
after the line "sack = dnf_sack_new ();"

However, there may be other applications in the world that use the libdnf
"context" and implement their own sack setup.

Related to: https://bugzilla.redhat.com/show_bug.cgi?id=1923820

jrohel committed:
[plugins] Add hooks to the dnf_goal_depsolve function

Added hooks:
PLUGIN_HOOK_ID_CONTEXT_PRE_GOAL_SOLVE
   Called just before the goal is resolved. The place where the plugin
   have a chance to change the goal.

PLUGIN_HOOK_ID_CONTEXT_GOAL_SOLVED
   Called shortly after the goal is resolved.

Added functions for accessing DnfPluginHookData delivered to plugins:
HyGoal hookContextGoalGetGoal(DnfPluginHookData * data);
DnfGoalActions * hookContextGoalGetActions(DnfPluginHookData * data);

Warning:
There was a missing context in the dnf_goal_depsolve function.
The code adds a pointer to DnfContext into the sack. The pointer is set
in the "dnf_context_setup_sack_with_flags" function. However, PackageKit
has it's own implementation of sack setup. So pointer to the context
will not be set in sack and the hooks will not be called.

In order for new hooks to be called with PackageKit, PackageKit must be
modified. -> Add a call to "dnf_sack_set_context (bag, job_data-> context);"
after the line "sack = dnf_sack_new ();"

However, there may be other applications in the world that use the libdnf
"context" and implement their own sack setup.

Conan-Kudo commented:
@jrohel The example plugin and documentation on that will need to be updated too.

jrohel commented:
When I joined the libdnf project, libdnf did not have plugin support. 
Due to an urgent request, I extended the context part of libdnf and added plugin support. However, the architecture of libdnf including the API was not plugin friendly. I solved where to put hooks in the existing code to make them useful. And how to modify the library without breaking existing applications. How to make plugins work with legacy applications.
Plugin support is functional and used. However, the solution contains compromises.

We have started developing a new library - libdnf5.  I have implemented plugin support in libdnf5 and it is already at a better level than in the old libdnf.

I will close this PR in favor of investing my time in the libdnf5 project.
